### ISSUE TITLE
Query Language

### ISSUE THREAD

**Issue by jenmud** [optiver/ruruki#1]
A query language needs to be implemented so that we are able to query the graph.

I was hoping that we could use [opencypher](http://www.opencypher.org/) , but seems that they are still far away being ready to use, and I am not sure what the API will need to look like in order to use it.

Any thoughts and ideas ?

**Comment by jenmud** [comment]
One idea was to use a [mongodb style query language](https://docs.mongodb.org/manual/tutorial/query-documents/). After looking at it, this sounds like a really good idea.

example:

```
 EntitySet.find(
    {
      $or: [
         { label: { $eq: "person" } },
         { label: { $eq: "cat" } }
       ],
       $and: [
            {name: {$contains: "awe"}}
        ],
    }
 )
```

I also like the idea that the `EntitySet` has a `.find` method that takes a query dictionary. I think that the `Graph` should support the same thing.

**Comment by afmacedo** [comment]
Good idea!!! :+1:

**Comment by jenmud** [comment]
It was you great idea, keep them coming.

**Comment by jenmud** [comment]
We have decided for now that only the graph will have a `.find` method on it for doing the queries. This is because we need to be able to construct queries including a mix of vertices and edges. And because the graph knows everything, it is best suited for the job.

**Comment by afmacedo** [comment]
First draft rolling in my fork. In comparison with Cyper it may get a little extensive (in it's current incarnation) though.
I.e.: 
> People who are "close friends" with John Doe

```python
    query = {
        '$heads': {
            '$and': [
                '$edges': {'label': 'close friend'},
                '$in_edges': {
                    '$filter': {'name': 'John Doe'}
                }
            ]
        }
    }
    GRAPH.find(query)
```

One can register variables for the subcommands by specifing ```|<varname>``` in front of a Query Operator. I.e.

```python
    query = {
        '$heads|cf': {
            '$and|cfe': [
                '$edges': {'label': 'close friend'},
                '$in_edges': {
                    '$filter|j': {'name': 'John Doe'}
                }
            ]
        }
    }
````

To register:
  * ```cf``` - all close friends with John
  * ```cfe``` - all _close friends_ edges connecting to _John_
  * ```j``` - _John_ himself

**Comment by jenmud** [comment]
I strongly support your ideas and implementations for the RUQL. It is a solid first step, though I think we can improve it and make it easier, I think you have done a stellar v1 here.

**Comment by goodtune** [comment]
The original example which was inspired by MongoDB's document query immediately reminded me of an LDAP query filter.

```
EntitySet.find(
    {
      $or: [
         { label: { $eq: "person" } },
         { label: { $eq: "cat" } }
       ],
       $and: [
            {name: {$contains: "awe"}}
        ],
    }
 )

LDAP.search("""
(&
    (|
        (label=person)
        (label=cat)
    )
    (name=*awe*)
)
""")
```

Given that there is already a grammar, existing parsers, what do you think of reusing an existing specification? Or is it not expressive enough?

**Comment by jenmud** [comment]
That is most definitely interesting and a possibility for starters. That depends on how easily we can add in custom functions for graph type expressions like **in_edges**, **out_edges**, **heads**, and so on.....

**Comment by jenmud** [comment]
# Just spitballing here

While the work that @afmacedo has been doing for the query language (RUQL), will be the query engine. I would really like to explore having a natural language parser like Cypher is ascii art.

So based on the [Gherkin](http://pythonhosted.org/behave/philosophy.html#the-gherkin-language) but modified. 

Functions which will be customizable and should be registered will be referred with a **$** in the front of it. Main actions/steps will be the first word in the sentence. For example, Find, And, Which, Then, etc... are default reserved actions/steps.


For example
```
Find me all vertices
    | label         | name |
    | PERSON | Bod    |
And properties $greater than
   | age |
   | 32   |
Which are connected by $outgoing edges
   | label       |
   | KNOWS |
Then return all $tail vertices as a $list
```
> So the above query would find all the **PERSON** labeled vertices that have a property name **Bob**, and is older than **32** and has a **KNOWS** relationship/edge to some other vertex. It will then return all the tail vertices in a list.


```
Find me all vertices
Which are connected to vertices
    | label    |
    | HOST  |
But only have 2 $hops
Then return the $path as a $list
```
> So this would find any vertices that are connected to a **HOST** vertex and only have a max of two hops from start to end. Then the result will be returned as a list of paths.

**Comment by techtonik** [comment]
Is there a language like this?

    find all children of (name='xxx') where (name='yyy')
    get age, label
    format csv

Or something line XPath?

**Comment by techtonik** [comment]
Or this stuff - http://graphql.org/ - looks nice.

**Comment by jenmud** [comment]
Agreed @techtonik GraphQL looks very interesting. I will read more about it, but I like what I see at the moment. Thanks for the suggestion.

As for your first comment, that is a is a nice way to ask the database questions. Seems pretty easy to read and write. I will look further into it.

At this stage we are still looking for what will be the best fit for a parser. The work that @afmacedo has been doing if quite effective and can do a lot already. But we need some sort of expression parser that will translate a query like your suggestion and convert it to the query structure that @afmacedo has been working on.

Nothing is set in stone though, as we looking for the easiest and best solution (as you always do).

Any other comments, feedback, or contributions will be greatly appreciated.

**Comment by techtonik** [comment]
It seems logical to add core support for plugging more query languages as modules. Especialy useful for cases when compatibility matters more than performance.

**Comment by jenmud** [comment]
After see a question in the openCypher project about an [in-memory graph database](https://github.com/opencypher/openCypher/issues/23) that supports cypher. It only validates that we are on the correct path, and implementing cypher as ruruki's query language is the right thing to do.

Full steam ahead!

**Comment by jenmud** [comment]
UPDATE

I have been reading, and looking into parsers. Seems that using [pyparser](https://pyparsing.wikispaces.com/) and creating ruruki grammar based on [openCypher EBNF](https://github.com/opencypher/openCypher/blob/master/grammar/generated/cypher.ebnf) is the way to go.

**Comment by techtonik** [comment]
Am I the only one who thinks that the name for this language should be `rucola`?

**Comment by jenmud** [comment]
I don't get it, what is it or does it mean?

I am first going to implement Cypher anyway.

**Comment by techtonik** [comment]
I don't know. 'ruruki uses rucola' is just a strange process of using it in a salad. Like most structured data is not a salad, but it starts to looks like this when you're trying to make sense of it.

**Comment by jenmud** [comment]
ruruki cypher parsing is going well so far. For the first iteration, we will be supporting MATCH and a simple return

eg
```
MATCH (n:PERSON {'name': 'Bob'})-[:KNOWS]->(f {'name': 'Fred'}) return n, f
```

**Comment by jenmud** [comment]
I have all the grammar parsing for doing matches complete (first draft anyways). Now that I have extracted the data, I just need to feed it into @afmacedo query engine.

@afmacedo, can you please let me know how far you are with the query engine and what branch can I used for testing.

**Comment by jenmud** [comment]
Loving it!! It is now all starting to come together

```python
In [18]: g.query("MATCH (a:person {name: 'peter'}), (b:person {'name__contains': 's'}) return a, b").vertices.all()
Out[18]: 
[<Vertex> ident: 5, label: person, properties: {u'age': 35, u'name': u'peter'},
 <Vertex> ident: 3, label: person, properties: {u'age': 32, u'name': u'josh'},
 <Vertex> ident: 1, label: person, properties: {u'age': 27, u'name': u'vadas'}]
```

**Comment by jenmud** [comment]
Now for the next setup in the query.

```
Match (n:Person) where n.name STARTS WITH "G" OR n.name STARTS WITH "D" return n;
```

So we are missing the parsing for the Predicates, `OR`, `AND`, `STARTS WITH`, `ENDS WITH`, `CONTAINS`, etc....

And need to add in functions too, like `filter`, `sum`, etc.....

**Comment by jenmud** [comment]
Whooo - completed the full opencypher parser thanks to my very knowledgeable college @OptiverTimAll . Now to start working in the execution plan, which is a whole new ball game.

**Comment by jenmud** [comment]
Found some good doco for the cypher execution plan http://neo4j.com/docs/stable/execution-plans.html
